docs/config.md
# Kumo

## Syntax

Kumo's configuration file uses the TOML format. The format's specification
can be found at _<https://toml.io/en/v1.0.0>_.

## Location

Kumo doesn't create the configuration file for you, but it looks for one at
<br> `${XDG_CONFIG_HOME:-$HOME/.config}/kumo/kumo.toml`.

## Fields

### font

This section documents the `[font]` table.

|Name|Description|Type|Default|
|-|-|-|-|
|family|Font family|text|`"sans"`|
|monospace_family|Monospace font family|text|`"mono"`|
|size|Font size|float|`16.0`|

### colors

This section documents the `[color]` table.

|Name|Description|Type|Default|
|-|-|-|-|
|foreground|Primary foreground color|color|`"#ffffff"`|
|background|Primary background color|color|`"#181818"`|
|highlight|Primary accent color|color|`"#752a2a"`|
|alt_foreground|Alternative foreground color|color|`"#bfbfbf"`|
|alt_background|Alternative background color|color|`"#282828"`|
|error|Error foreground color|color|`"#ac4242"`|
|disabled|Disabled foreground color|color|`"#666666"`|
|dark_mode|Whether pages should use a dark theme if available|boolean|`true`|

### search

This section documents the `[input]` table.

|Name|Description|Type|Default|
|-|-|-|-|
|uri|Search engine URI.<br><br>The search query will be appended to the end of this URI.|text|`"https://duckduckgo.com/?q="`|

### input

This section documents the `[input]` table.

|Name|Description|Type|Default|
|-|-|-|-|
|max_tap_distance|Square of the maximum distance before touch input is considered a drag|float|`400.0`|
|max_multi_tap|Maximum interval between taps to be considered a double/trible-tap|integer (milliseconds)|`300`|
|long_press|Minimum time before a tap is considered a long-press|integer (milliseconds)|`300`|
|velocity_interval|Milliseconds per velocity tick|integer|`30`|
|velocity_friction|Percentage of velocity retained each tick|float|`0.85`|

### engine

This section documents the `[engine]` table.

|Name|Description|Type|Default|
|-|-|-|-|
|default|Default browser engine|"WebKit" \| "Servo"|`WebKit`|
|max_background_tabs|Maximum number of loaded invisible tabs.<br><br>If the number of background tabs exceeds this value, the least recently used tabs will be unloaded.<br><br>This limit is per window.|integer|`3`|


src/config.rs
//! Configuration options.

use std::fmt::{self, Display, Formatter};
use std::ops::Deref;
use std::sync::{Arc, LazyLock, RwLock};
use std::time::Duration;

use configory::docgen::{DocType, Docgen, Leaf};
use configory::{EventHandler, Manager};
use funq::MtQueueHandle;
use serde::de::Visitor;
use serde::{Deserialize, Deserializer};
use tracing::{error, info};

use crate::engine::EngineType;
use crate::{Error, State};

#[funq::callbacks(State)]
trait ConfigHandler {
    /// Config reload handler.
    fn config_reloaded(&mut self);
}

impl ConfigHandler for State {
    fn config_reloaded(&mut self) {
        for window in self.windows.values_mut() {
            // Update settings UI.
            window.reload_settings();

            // Trigger redraw if rendering is stalled.
            window.unstall();
        }
    }
}

/// Shared configuration state.
pub static CONFIG: LazyLock<Arc<RwLock<Config>>> =
    LazyLock::new(|| Arc::new(RwLock::new(Config::default())));

/// Initialize configuration state.
pub fn init_config(queue: MtQueueHandle<State>) -> Result<Manager<ConfigEventHandler>, Error> {
    let config_manager = Manager::new("kumo", ConfigEventHandler::new(queue))?;

    // Load initial configuration.
    let config = config_manager
        .get::<&str, Config>(&[])
        .inspect_err(|err| error!("Config error: {err}"))
        .ok()
        .flatten()
        .unwrap_or_default();
    *CONFIG.write().unwrap() = config;

    Ok(config_manager)
}

/// # Kumo
///
/// ## Syntax
///
/// Kumo's configuration file uses the TOML format. The format's specification
/// can be found at _<https://toml.io/en/v1.0.0>_.
///
/// ## Location
///
/// Kumo doesn't create the configuration file for you, but it looks for one at
/// <br> `${XDG_CONFIG_HOME:-$HOME/.config}/kumo/kumo.toml`.
///
/// ## Fields
#[derive(Docgen, Deserialize, Default, Debug)]
#[serde(default, deny_unknown_fields)]
pub struct Config {
    /// This section documents the `[font]` table.
    pub font: Font,
    /// This section documents the `[color]` table.
    pub colors: Colors,
    /// This section documents the `[input]` table.
    pub search: Search,
    /// This section documents the `[input]` table.
    pub input: Input,
    /// This section documents the `[engine]` table.
    pub engine: Engine,

    /// Incremental config ID, to track changes.
    #[serde(skip)]
    #[docgen(skip)]
    pub generation: u32,
}

/// Font configuration.
#[derive(Docgen, Deserialize, Debug)]
#[serde(default, deny_unknown_fields)]
pub struct Font {
    /// Font family.
    pub family: FontFamily,
    /// Monospace font family.
    pub monospace_family: FontFamily,
    /// Font size.
    pub size: f64,
}

impl Default for Font {
    fn default() -> Self {
        Self {
            monospace_family: FontFamily::from("mono"),
            family: FontFamily::from("sans"),
            size: 16.,
        }
    }
}

impl Font {
    /// Get font size relative to the default.
    pub fn size(&self, scale: f64) -> u8 {
        (self.size * scale).round() as u8
    }
}

/// Color configuration.
#[derive(Docgen, Deserialize, Copy, Clone, Hash, PartialEq, Eq, Debug)]
#[serde(default, deny_unknown_fields)]
pub struct Colors {
    /// Primary foreground color.
    #[serde(alias = "fg")]
    pub foreground: Color,
    /// Primary background color.
    #[serde(alias = "bg")]
    pub background: Color,
    /// Primary accent color.
    #[serde(alias = "hl")]
    pub highlight: Color,

    /// Alternative foreground color.
    #[serde(alias = "alt_fg")]
    pub alt_foreground: Color,
    /// Alternative background color.
    #[serde(alias = "alt_bg")]
    pub alt_background: Color,

    /// Error foreground color.
    pub error: Color,
    /// Disabled foreground color.
    pub disabled: Color,

    /// Whether pages should use a dark theme if available.
    pub dark_mode: bool,
}

impl Default for Colors {
    fn default() -> Self {
        Self {
            foreground: Color::new(255, 255, 255),
            background: Color::new(24, 24, 24),
            highlight: Color::new(117, 42, 42),

            alt_foreground: Color::new(191, 191, 191),
            alt_background: Color::new(40, 40, 40),

            error: Color::new(172, 66, 66),
            disabled: Color::new(102, 102, 102),

            dark_mode: true,
        }
    }
}

#[derive(Docgen, Deserialize, Debug)]
#[serde(default, deny_unknown_fields)]
pub struct Search {
    /// Search engine URI.
    ///
    /// The search query will be appended to the end of this URI.
    pub uri: String,
}

impl Default for Search {
    fn default() -> Self {
        Self { uri: "https://duckduckgo.com/?q=".into() }
    }
}

/// Input configuration.
#[derive(Docgen, Deserialize, Debug)]
#[serde(default, deny_unknown_fields)]
pub struct Input {
    /// Square of the maximum distance before touch input is considered a drag.
    pub max_tap_distance: f64,
    /// Maximum interval between taps to be considered a double/trible-tap.
    #[docgen(doc_type = "integer (milliseconds)", default = "300")]
    pub max_multi_tap: MillisDuration,
    /// Minimum time before a tap is considered a long-press.
    #[docgen(doc_type = "integer (milliseconds)", default = "300")]
    pub long_press: MillisDuration,

    /// Milliseconds per velocity tick.
    pub velocity_interval: u16,
    /// Percentage of velocity retained each tick.
    pub velocity_friction: f64,
}

impl Default for Input {
    fn default() -> Self {
        Self {
            max_multi_tap: Duration::from_millis(300).into(),
            long_press: Duration::from_millis(300).into(),
            velocity_interval: 30,
            velocity_friction: 0.85,
            max_tap_distance: 400.,
        }
    }
}

#[derive(Docgen, Deserialize, Debug)]
#[serde(default, deny_unknown_fields)]
pub struct Engine {
    /// Default browser engine.
    pub default: EngineType,
    /// Maximum number of loaded invisible tabs.
    ///
    /// If the number of background tabs exceeds this value, the least recently
    /// used tabs will be unloaded.
    ///
    /// This limit is per window.
    pub max_background_tabs: usize,
}

impl Default for Engine {
    fn default() -> Self {
        Self {
            max_background_tabs: 3,
            #[cfg(feature = "webkit")]
            default: EngineType::WebKit,
            #[cfg(not(feature = "webkit"))]
            default: EngineType::Servo,
        }
    }
}

/// RGB color.
#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug)]
pub struct Color {
    pub r: u8,
    pub g: u8,
    pub b: u8,
}

impl Color {
    pub const fn new(r: u8, g: u8, b: u8) -> Self {
        Self { r, g, b }
    }

    pub const fn as_u8(&self) -> [u8; 4] {
        [self.r, self.g, self.b, 255]
    }

    pub const fn as_u16(&self) -> [u16; 3] {
        let factor = u16::MAX / u8::MAX as u16;
        [self.r as u16 * factor, self.g as u16 * factor, self.b as u16 * factor]
    }

    pub const fn as_u32(&self) -> [u32; 3] {
        let factor = u32::MAX / u8::MAX as u32;
        [self.r as u32 * factor, self.g as u32 * factor, self.b as u32 * factor]
    }

    pub const fn as_f32(&self) -> [f32; 3] {
        [self.r as f32 / 255., self.g as f32 / 255., self.b as f32 / 255.]
    }

    pub const fn as_f64(&self) -> [f64; 3] {
        [self.r as f64 / 255., self.g as f64 / 255., self.b as f64 / 255.]
    }
}

impl Docgen for Color {
    fn doc_type() -> DocType {
        DocType::Leaf(Leaf::new("color"))
    }

    fn format(&self) -> String {
        format!("\"#{:0>2x}{:0>2x}{:0>2x}\"", self.r, self.g, self.b)
    }
}

/// Deserialize rgb color from a hex string.
impl<'de> Deserialize<'de> for Color {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct ColorVisitor;

        impl Visitor<'_> for ColorVisitor {
            type Value = Color;

            fn expecting(&self, f: &mut Formatter<'_>) -> fmt::Result {
                f.write_str("hex color like #ff00ff")
            }

            fn visit_str<E>(self, value: &str) -> Result<Color, E>
            where
                E: serde::de::Error,
            {
                let channels = match value.strip_prefix('#') {
                    Some(channels) => channels,
                    None => {
                        return Err(E::custom(format!("color {value:?} is missing leading '#'")));
                    },
                };

                let digits = channels.len();
                if digits != 6 {
                    let msg = format!("color {value:?} has {digits} digits; expected 6");
                    return Err(E::custom(msg));
                }

                match u32::from_str_radix(channels, 16) {
                    Ok(mut color) => {
                        let b = (color & 0xFF) as u8;
                        color >>= 8;
                        let g = (color & 0xFF) as u8;
                        color >>= 8;
                        let r = color as u8;

                        Ok(Color::new(r, g, b))
                    },
                    Err(_) => Err(E::custom(format!("color {value:?} contains non-hex digits"))),
                }
            }
        }

        deserializer.deserialize_str(ColorVisitor)
    }
}

impl Display for Color {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
        write!(f, "#{:0>2x}{:0>2x}{:0>2x}", self.r, self.g, self.b)
    }
}

/// Config wrapper for millisecond-precision durations.
#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug)]
pub struct MillisDuration(Duration);

impl Deref for MillisDuration {
    type Target = Duration;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'de> Deserialize<'de> for MillisDuration {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let ms = u64::deserialize(deserializer)?;
        Ok(Duration::from_millis(ms).into())
    }
}

impl From<Duration> for MillisDuration {
    fn from(duration: Duration) -> Self {
        Self(duration)
    }
}

impl Display for MillisDuration {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
        write!(f, "{}", self.0.as_millis())
    }
}

/// Read-only reference counted font family string.
#[derive(Hash, PartialEq, Eq, Clone, Debug)]
pub struct FontFamily {
    arc: Arc<String>,
}

impl FontFamily {
    fn from(family: impl Into<String>) -> Self {
        Self { arc: Arc::new(family.into()) }
    }
}

impl Deref for FontFamily {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        self.arc.as_str()
    }
}

impl Display for FontFamily {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
        write!(f, "{}", self.arc.as_str())
    }
}

impl<'de> Deserialize<'de> for FontFamily {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let family = String::deserialize(deserializer)?;
        Ok(Self::from(family))
    }
}

impl Docgen for FontFamily {
    fn doc_type() -> DocType {
        String::doc_type()
    }

    fn format(&self) -> String {
        self.arc.format()
    }
}

/// Event handler for configuration manager updates.
pub struct ConfigEventHandler {
    queue: MtQueueHandle<State>,
}

impl ConfigEventHandler {
    fn new(queue: MtQueueHandle<State>) -> Self {
        Self { queue }
    }

    /// Reload config and unstall renderer.
    fn reload_config(&self, config: &configory::Config) {
        reload_config(config);
        self.queue.clone().config_reloaded();
    }
}

impl EventHandler for ConfigEventHandler {
    type MessageData = ();

    fn file_changed(&self, config: &configory::Config) {
        self.reload_config(config);
    }

    fn ipc_changed(&self, config: &configory::Config) {
        self.reload_config(config);
    }

    fn file_error(&self, _config: &configory::Config, err: configory::Error) {
        error!("Configuration file error: {err}");
    }
}

/// Update the configuration file.
pub fn reload_config(config: &configory::Config) {
    info!("Reloading configuration file");

    // Parse config or fall back to the default.
    let parsed = config
        .get::<&str, Config>(&[])
        .inspect_err(|err| error!("Config error: {err}"))
        .ok()
        .flatten()
        .unwrap_or_default();

    // Calculate generation based on current config.
    let mut config = CONFIG.write().unwrap();
    let next_generation = config.generation + 1;

    // Update the config.
    *config = parsed;
    config.generation = next_generation;
}

#[cfg(all(test, feature = "webkit"))]
mod tests {
    use std::fs;

    use configory::docgen::markdown::Markdown;

    use super::*;

    #[test]
    fn config_docs() {
        let mut formatter = Markdown::new();
        formatter.set_heading_size(3);
        let expected = formatter.format::<Config>();

        // Uncomment to update config documentation.
        // fs::write("./docs/config.md", &expected).unwrap();

        // Ensure documentation is up to date.
        let docs = fs::read_to_string("./docs/config.md").unwrap();
        assert_eq!(docs, expected);
    }
}
